torch/_inductor/lowering.py

```diff
@@ -3912,20 +3926,115 @@ def fn(idx, return_index):
         else:
             return maxval
 
-    r1 = Pointwise.create(
+    out = Pointwise.create(
         device=x.get_device(),
         dtype=x.get_dtype(),
         inner_fn=functools.partial(fn, return_index=False),
         ranges=new_size,
     )
-    r2 = Pointwise.create(
+    offsets = Pointwise.create(
         device=x.get_device(),
-        dtype=torch.int64,
+        dtype=offset_dtype,
         inner_fn=functools.partial(fn, return_index=True),
         ranges=new_size,
     )
-    # TODO(jansel): should we force these to be realized?
-    return r1, r2
+    return out, offsets
+
+
+@register_lowering(prims._low_memory_max_pool2d_with_offsets, type_promotion_kind=None)
+def _low_memory_max_pool2d_with_offsets(
+    x, kernel_size, stride, padding, dilation, ceil_mode, *, offset_dtype
+):
+    # assert we are not on a fallback path, the inductor decomp should have guaranteed this
+    kernel_size, stride, padding, dilation, _ = max_pool2d_checks(
+        x, kernel_size, stride, padding, dilation, assert_fallback=False
+    )
+    return max_pool2d_with_offsets_impl(
+        x, kernel_size, stride, padding, ceil_mode=ceil_mode, offset_dtype=offset_dtype
+    )
+
+
+@register_lowering(aten.max_pool2d_with_indices, type_promotion_kind=None)
+def max_pool2d_with_indices(
+    x, kernel_size, stride=None, padding=0, dilation=1, ceil_mode=False
+):
+    kernel_size, stride, padding, dilation, use_fallback = max_pool2d_checks(
+        x, kernel_size, stride, padding, dilation
+    )
+    if use_fallback:
+        return fallback_max_pool2d_with_indices(
+            x, kernel_size, stride, padding, dilation, ceil_mode
+        )
+
+    out, offsets = max_pool2d_with_offsets_impl(
+        x, kernel_size, stride, padding, ceil_mode=ceil_mode, offset_dtype=torch.int32
+    )
+    indices = max_pool2d_offsets_to_indices_impl(
+        offsets, kernel_size[1], x.get_size()[-1], stride, padding
+    )
+    # If we landed here, the indices are not used for the backward pass, realize
+    # hint here so the conversion is fused into the forward loop.
+    indices.realize_hint()
+    return out, indices
+
+
+def max_pool2d_offsets_to_indices_impl(
+    offsets, kernel_width, input_width, stride, padding
+):
+    # TODO: Generalize to other max pooling flavors, and arbitrary dim
+
+    offsets_loader = offsets.make_loader()
+
+    def get_input_width_opsvalue():
+        # input width usually comes from a symbolic sizevar on the input
+        if isinstance(input_width, sympy.Expr):
+            assert input_width.is_integer
+            return ops.index_expr(input_width, torch.int64)
+        else:
+            # routing through adaptive_max_pool to here the shape is evaluated
+            # so we get a constant
+            return ops.constant(input_width, torch.int64)
```

Comment:
Hmm, `routing through adaptive_max_pool to here the shape is evaluated` 

Hmm, what do you mean ?

Reply:
I was trying to explain why a branch was needed there, as looking at the paths through  `max_pool2d_with_indices` it seems it should not be needed.  Indeed we don't hit this branch though any of the `max_pool2d` tests. 


`adaptive_max_pool2d` will route through here under [special conditions](https://github.com/pytorch/pytorch/blob/c63a7b569133c9d91bde362c68e4f60abd4b619b/torch/_inductor/lowering.py#L4305C4-L4307C55). In that context the lowering for `adaptive_max_pool` has called `evaluate_static_shape` on the input sizes and `input_width` is always a constant.